ledger/error/src/main/scala/com/daml/error/ErrorCode.scala

```diff
@@ -0,0 +1,114 @@
+  *
+  * Errors are organised according to ErrorGroups. And we separate the error code definition (using a singleton
+  * by virtue of using objects to express the nested hierarchy) and the actual error.
+  *
+  * Please note that there is some implicit argument passing involved in the example below:
+  *
+  * object SyncServiceErrors extends ParticipantErrorGroup {
+  *   object ConnectionErrors extends ErrorGroup {
+  *     object DomainUnavailable extends ErrorCode(id="DOMAIN_UNAVAILABLE", ..) {
+  *        case class ActualError(someContext: Val) extends BaseError with SyncServiceError
+  *        // this error will actually be referring to the same error code!
+  *        case class OtherError(otherContext: Val) extends BaseError with SyncServiceError
+  *     }
+  *   }
+  *   object HandshakeErrors extends ErrorGroup {
+  *     ...
+  *   }
+  * }
+  */
+abstract class ErrorCode(val id: String, val category: ErrorCategory)(implicit
+    val parent: ErrorClass
+) {
+
+  // TODO error codes: Implement `asGrpcError`
+
+  require(id.nonEmpty, "error-id must be non empty")
+  require(id.length < 64, s"error-id is too long: $id")
+  require(id.forall(c => c.isUpper || c == '_' || c.isDigit), s"Invalid characters in error-id $id")
+
+  implicit val code: ErrorCode = this
+
+  def toMsg(cause: => String): String =
+    s"${ErrorCode.truncateCause(cause)}"
+
+  // TODO error codes: Pass correlation id
+  // TODO error codes: Allow injectable tracing
+  def log(logger: ContextualizedLogger, err: BaseError)(implicit
+      loggingContext: LoggingContext
+  ): Unit = {
+    val message = toMsg(err.cause)
+    // TODO error codes: Enrich logging context with:
+    //                     * error context
+    //                     * error location
+    //                     * error code
+    (logLevel, err.throwableO) match {
+      case (Level.INFO, None) => logger.info(message)
+      case (Level.INFO, Some(tr)) => logger.info(message, tr)
+      case (Level.WARN, None) => logger.warn(message)
+      case (Level.WARN, Some(tr)) => logger.warn(message, tr)
+      // an error that is logged with < INFO is not an error ...
+      case (_, None) => logger.error(message)
+      case (_, Some(tr)) => logger.error(message, tr)
+    }
+  }
+
+  /** log level of the error code
+    *
+    * Generally, the log level is defined by the error category. In rare cases, it might be overridden
+    * by the error code.
+    */
+  protected def logLevel: Level = category.logLevel
+
+  /** True if this error may appear on the API */
+  protected def exposedViaApi: Boolean = category.grpcCode.nonEmpty
+
+  /** The error conveyance doc string provides a statement about the form this error will be returned to the user */
+  private[error] def errorConveyanceDocString: Option[String] = {
+    val loggedAs = s"This error is logged with log-level $logLevel on the server side."
+    val apiLevel = (category.grpcCode, exposedViaApi) match {
+      case (Some(code), true) =>
```

Comment:
To improve readability, I'd suggest branching on `securitySensitive` as part of the pattern match and adding the newline separately, e.g.:
```scala
  val apiLevel = ... match {
    case (Some(code), true) if category.securitySensitive =>
      s"This error is exposed ..."
    case (Some(code), true) if !category.securitySensitive =>
      s"This error is exposed ..."
    case _ => ""
  }
  ...
```